bin/client/src/precompiles/bls12.rs

```diff
@@ -0,0 +1,108 @@
+const PAIRING_MULTIPLIER_BASE: u64 = 32600;
+
+/// Offset gas fee for BLS12-381 pairing operation.
+const PAIRING_OFFSET_BASE: u64 = 37700;
+
+/// The address of the BLS12-381 pairing precompile.
+pub(crate) const FPVM_BLS12_PAIRING: PrecompileWithAddress =
+    PrecompileWithAddress(BLS12_PAIRING_CHECK, Precompile::Standard(fpvm_bls12_pairing));
+
+/// Performs an FPVM-accelerated BLS12-381 pairing check.
+fn fpvm_bls12_pairing(input: &Bytes, gas_limit: u64) -> PrecompileResult {
+    let input_len = input.len();
+    if input_len % INPUT_LENGTH != 0 {
+        return Err(PrecompileError::Other(alloc::format!(
+            "Pairing input length should be multiple of {INPUT_LENGTH}, was {input_len}"
+        ))
+        .into());
+    }
+
+    let k = input_len / INPUT_LENGTH;
+    let required_gas: u64 = PAIRING_MULTIPLIER_BASE * k as u64 + PAIRING_OFFSET_BASE;
+    if required_gas > gas_limit {
+        return Err(PrecompileError::OutOfGas.into());
+    }
+
+    let result_data = kona_proof::block_on(async move {
+        // Write the hint for the ecrecover precompile run.
+        let hint_data = &[BLS12_PAIRING_CHECK.as_ref(), input.as_ref()];
+        HINT_WRITER
+            .write(&HintType::L1Precompile.encode_with(hint_data))
+            .await
+            .map_err(OracleProviderError::Preimage)?;
+
+        // Construct the key hash for the ecrecover precompile run.
+        let raw_key_data = hint_data.iter().copied().flatten().copied().collect::<Vec<u8>>();
+        let key_hash = keccak256(&raw_key_data);
+
+        // Fetch the result of the ecrecover precompile run from the host.
+        let result_data = ORACLE_READER
+            .get(PreimageKey::new(*key_hash, PreimageKeyType::Precompile))
+            .await
+            .map_err(OracleProviderError::Preimage)?;
+
+        // Ensure we've received valid result data.
+        if result_data.is_empty() {
+            return Err(OracleProviderError::Preimage(PreimageOracleError::Other(
+                "Invalid result data".to_string(),
+            )));
+        }
+
+        // Ensure we've not received an error from the host.
+        if result_data[0] == 0 {
+            return Err(OracleProviderError::Preimage(PreimageOracleError::Other(
+                "Error executing ecrecover precompile in host".to_string(),
+            )));
+        }
+
+        // Return the result data.
+        Ok(result_data[1..].to_vec())
+    })
+    .map_err(|e| PrecompileError::Other(e.to_string()))?;
+
+    Ok(PrecompileOutput::new(required_gas, result_data.into()))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use alloc::vec;
+
+    #[test]
+    fn test_fpvm_bls12_out_of_gas() {
+        let input = Bytes::from(vec![0u8; INPUT_LENGTH * 2]);
+        let gas_limit = PAIRING_MULTIPLIER_BASE - 1;
+        assert_eq!(
+            fpvm_bls12_pairing(&input, gas_limit),
+            Err(PrecompileError::OutOfGas.into())
+        );
+    }
```

Comment:
can we get tests for all the possible errors returned from `fpvm_bls12_pairing`?

Reply:
Added another test for the offset. Note that other precompiles do not have tests. Can you write up a ticket for adding tests to those precompiles?

Reply:
believe @meyer9 has been testing this elsewhere?

Reply:
Yeah I copied some test cases from here for the end-to-end tests in the Optimism monorepo: https://eips.ethereum.org/assets/eip-2537/test-vectors

I would think that revm should include the test vectors from here as part of the tests for those precompiles, but I don't see them. What do you think about adding these to revm?

Reply:
that's a great idea! do you want to open an issue in revm? think it makes sense if you open it since you have the most context here

Reply:
sure, will do!

Reply:
https://github.com/bluealloy/revm/issues/2034

Reply:
ah nvm, those tests are already implemented as part of the `execution-spec-tests` repo. https://github.com/bluealloy/revm/blob/4c773541ddf5a4966a512b4f0dde3da617ba9962/tests/pectra_devnet5/state_tests/prague/eip2537_bls_12_381_precompiles/bls12_g1msm/valid.json#L1